heka/plugins/kafkaconsumergroup/kafka_consumer_group_input.go

```diff
@@ -0,0 +1,272 @@
+	// contains useful information for debugging consumer group partition
+	// changes
+	if k.config.LogSarama {
+		sarama.Logger = log.New(os.Stdout, "[Sarama] ", log.LstdFlags)
+	}
+
+	k.consumerConfig = consumergroup.NewConfig()
+	switch k.config.OffsetMethod {
+	case "Newest":
+		k.consumerConfig.Offsets.Initial = sarama.OffsetNewest
+	case "Oldest":
+		k.consumerConfig.Offsets.Initial = sarama.OffsetOldest
+	default:
+		return fmt.Errorf("invalid offset_method: %s", k.config.OffsetMethod)
+	}
+
+	k.consumerConfig.Offsets.ProcessingTimeout = 10 * time.Second
+
+	k.consumerConfig.Config.Metadata.Retry.Max = k.config.MetadataRetries
+	k.consumerConfig.Config.Metadata.Retry.Backoff = time.Duration(k.config.WaitForElection) * time.Millisecond
+	k.consumerConfig.Config.Metadata.RefreshFrequency = time.Duration(k.config.BackgroundRefreshFrequency) * time.Millisecond
+
+	k.consumerConfig.Config.Net.MaxOpenRequests = k.config.MaxOpenRequests
+	k.consumerConfig.Config.Net.DialTimeout = time.Duration(k.config.DialTimeout) * time.Millisecond
+	k.consumerConfig.Config.Net.ReadTimeout = time.Duration(k.config.ReadTimeout) * time.Millisecond
+	k.consumerConfig.Config.Net.WriteTimeout = time.Duration(k.config.WriteTimeout) * time.Millisecond
+
+	k.consumerConfig.Config.Consumer.Fetch.Default = k.config.DefaultFetchSize
+	k.consumerConfig.Config.Consumer.Fetch.Min = k.config.MinFetchSize
+	k.consumerConfig.Config.Consumer.Fetch.Max = k.config.MaxMessageSize
+	k.consumerConfig.Config.Consumer.MaxWaitTime = time.Duration(k.config.MaxWaitTime) * time.Millisecond
+	k.consumerConfig.Config.ChannelBufferSize = k.config.EventBufferSize
+
+	var zookeeperNodes []string
+	zookeeperNodes, k.consumerConfig.Zookeeper.Chroot = kazoo.ParseConnectionString(k.config.ZookeeperConnectionString)
+	if len(zookeeperNodes) == 0 {
+		return fmt.Errorf("unable to parse zookeeper_connection_string")
+	}
+
+	consumer, err := consumergroup.JoinConsumerGroup(k.config.ConsumerGroup, k.config.Topics, zookeeperNodes, k.consumerConfig)
+	if err != nil {
+		return
+	}
+	k.consumer = consumer
+	k.stopChan = make(chan bool)
+	return
+}
+
+func (k *KafkaConsumerGroupInput) addField(pack *pipeline.PipelinePack, name string,
+	value interface{}, representation string) {
+
+	if field, err := message.NewField(name, value, representation); err == nil {
+		pack.Message.AddField(field)
+	} else {
+		k.ir.LogError(fmt.Errorf("can't add '%s' field: %s", name, err.Error()))
+	}
+}
+
+func (k *KafkaConsumerGroupInput) Run(ir pipeline.InputRunner, h pipeline.PluginHelper) (err error) {
+	sRunner := ir.NewSplitterRunner("")
+
+	defer func() {
+		if err := k.consumer.Close(); err != nil {
+			k.ir.LogError(fmt.Errorf("error closing the consumer: %s", err.Error()))
+		}
+		sRunner.Done()
+	}()
+	k.ir = ir
+
+	go func() {
+		for err := range k.consumer.Errors() {
+			// this isn't necessarily a process message failure, as the failure
+			// channel is async
+			// atomic.AddInt64(&k.processMessageFailures, 1)
```

Comment:
mmm, the error usually is a process_message error (just not for the current message) correct?  I would increment it